torchtitan/train.py

```diff
@@ -428,14 +487,20 @@ def train(self):
                 sync_every=job_config.fault_tolerance.sync_steps,
             ),
         ):
-            for inputs, labels in self.batch_generator(self.dataloader):
-                if self.step >= job_config.training.steps:
-                    break
+            data_iterator = self.batch_generator(self.dataloader)
+            data_ran_out = False
+            while self.step < job_config.training.steps and not data_ran_out:
```

Comment:
With the above proposal, we only need to wrap `self.step += 1` and `self.gc_handler.run(self.step)` with `self.begin_step()`. Else just increase `curr_accumulation_step` .